pkg/controller/direct/sql/normalize.go

```diff
@@ -0,0 +1,307 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package sql
+
+import (
+	"context"
+	"fmt"
+
+	refs "github.com/GoogleCloudPlatform/k8s-config-connector/apis/refs/v1beta1"
+	krm "github.com/GoogleCloudPlatform/k8s-config-connector/apis/sql/v1beta1"
+	kmsv1beta1 "github.com/GoogleCloudPlatform/k8s-config-connector/pkg/clients/generated/apis/kms/v1beta1"
+	storagev1beta1 "github.com/GoogleCloudPlatform/k8s-config-connector/pkg/clients/generated/apis/storage/v1beta1"
+	"github.com/GoogleCloudPlatform/k8s-config-connector/pkg/controller/direct"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func NormalizeSQLInstance(ctx context.Context, kube client.Reader, obj *krm.SQLInstance) error {
+	if err := normalizeCryptoKeyRef(ctx, kube, obj); err != nil {
+		return err
+	}
+	if err := normalizeMasterInstanceRef(ctx, kube, obj); err != nil {
+		return err
+	}
+	if err := normalizeReplicaPasswordRef(ctx, kube, obj); err != nil {
+		return err
+	}
+	if err := normalizeRootPasswordRef(ctx, kube, obj); err != nil {
+		return err
+	}
+	if err := normalizePrivateNetworkRef(ctx, kube, obj); err != nil {
+		return err
+	}
+	if err := normalizeAuditLogBucketRef(ctx, kube, obj); err != nil {
+		return err
+	}
+	if err := normalizeTFDefaults(obj); err != nil {
+		return err
+	}
+	return nil
+}
+
+func normalizeCryptoKeyRef(ctx context.Context, kube client.Reader, obj *krm.SQLInstance) error {
+	if obj.Spec.EncryptionKMSCryptoKeyRef == nil {
+		return nil
+	}
+
+	keyRef := obj.Spec.EncryptionKMSCryptoKeyRef
+
+	if keyRef.Kind != "" && keyRef.Kind != "KMSCryptoKey" {
+		return fmt.Errorf("spec.encryptionKMSCryptoKeyRef.kind must be KMSCryptoKey")
+	}
+
+	if keyRef.External != "" && keyRef.Name != "" {
+		return fmt.Errorf("cannot specify both spec.encryptionKMSCryptoKeyRef.external and spec.encryptionKMSCryptoKeyRef.name")
+	}
+
+	if keyRef.External != "" {
```

Comment:
Why? I felt like it made more sense to use name, because it felt like bringing the "external info" to the "internal". Maybe somewhat of a personal preference thing. Alternatively, perhaps we shouldn't re-use the CR fields, and instead store the resolved references in separate fields within the adapter. Wdyt?

Reply:
IMO, `name` here strictly means the K8s CR name of the referenced resource. It doesn't make sense to expend it to other values. And `external` here means the value we will eventually pass to the external system (the GCP API). But I understand different people have different takes on it. So not reusing the field sounds like a cleaner/easier solution.

Reply:
ok, I put the refs in an field in the adapter